File: 1094-car-pooling/1094-car-pooling.cpp
class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        map<int,int>mp;
        for(auto &it: trips){
            mp[it[1]] += it[0];
            mp[it[2]] -= it[0];
        }

        int count = 0;
        for(auto &it : mp) {
            count += it.second;
            if(count > capacity) {
                return false;
            }
        }
        return true;
    }
};

File: 1094-car-pooling/1094-car-pooling.py
from typing import List

class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        mp = {}

        # difference map
        for passengers, start, end in trips:
            mp[start] = mp.get(start, 0) + passengers
            mp[end]   = mp.get(end, 0)   - passengers

        curr = 0

        # sweep timeline
        for _, val in sorted(mp.items()):
            curr += val
            if curr > capacity:
                return False

        return True
